restapi/auth.go

```diff
@@ -33,47 +34,29 @@ func generateStateOauthCookie(w http.ResponseWriter) (string, error) {
 	return state, nil
 }
 
-func (api api) Login(w http.ResponseWriter, r *http.Request) render.Renderer {
-	state, err := generateStateOauthCookie(w)
-	if err != nil {
-		return rest.ErrInternal(api.logger, err)
-	}
-
-	url := api.oauthconfig.AuthCodeURL(state)
-	http.Redirect(w, r, url, http.StatusTemporaryRedirect)
-
-	return rest.JSONStatusOK(url)
+type GoogleClaims struct {
+	Email string `mapstructure:"email"`
+	Hd    string `mapstructure:"hd"`
 }
 
-func (api api) AuthCallback(w http.ResponseWriter, r *http.Request) render.Renderer {
-	expectedState, _ := r.Cookie("oauthstate")
-	if expectedState.Value != r.FormValue("state") {
-		return rest.ErrUnauthorized("Invalid oauth state")
-	}
+func (api api) Login(w http.ResponseWriter, r *http.Request) render.Renderer {
 
-	// exchange received authorization code for an access token
-	token, err := api.oauthconfig.Exchange(context.Background(), r.FormValue("code"))
+	payload, err := idtoken.Validate(oauth2.NoContext, r.Header.Get("Authorization"), config.GetGoogleClientId())
 	if err != nil {
-		return rest.ErrUnauthorized("Invalid authorization code")
+		return rest.ErrInternal(api.logger, err)
 	}
 
-	// use access token to create a client which we use to access user info
-	user := models.User{}
-	client := api.oauthconfig.Client(context.Background(), token)
-	response, err := client.Get("https://www.googleapis.com/oauth2/v3/userinfo")
-	if err != nil {
-		return rest.JSONStatusOK("Invalid access token")
-	}
-	defer response.Body.Close()
-	err = json.NewDecoder(response.Body).Decode(&user)
-	if err != nil {
-		return rest.ErrInternal(api.logger, err)
+	googleClaims := GoogleClaims{}
+	mapstructure.Decode(payload.Claims, &googleClaims)
+	user := models.User{
+		Email: googleClaims.Email,
+		Hd: googleClaims.Hd,
 	}
 
-	// redirect if invalid user
+	// create user in database if valid
 	err = api.database.FirstOrCreate(&user, models.User{Email: user.Email}).Error
 	if err != nil {
-		http.Redirect(w, r, "/unauthorized", http.StatusTemporaryRedirect)
+		return rest.ErrInternal(api.logger, err)
```

Comment:
Right now its in the cookie but i wouldn't be opposed to just returning it and letting the frontend store it however they want

Reply:
iirc we chose to store it in a cookie so that fe wouldn't need to worry about storage and retrieving & sending the token as part of the request (i.e. setting it in the req header every time)